### DISCUSSION TITLE
implementing a resource check: how to specify metadata in json?

### DISCUSSION THREAD

**Discussion by x3nb63** [concourse/concourse#6472] [Help & Support] [answered]
I am implementing my own resource and want the `check` it to show metadata in the WebUI, just like the git-resource does.

But all the WebUI gives me is json errors, such as this:

```
run check step: run check step: check: json: cannot unmarshal object into Go value of type []atc.Version

when parsing resource response:

{
  "version": {
    "value": "a version string"
  },
  "metadata": [
    {
      "name": "somename",
      "value": "somevalue"
    },
    {
      "name": "othername",
      "value": "othervalue"
    }
  ]
}
```
trying with the json shown [here](https://concourse-ci.org/implementing-resource-types.html) for the `in` method gives the same error:

```
run check step: run check step: check: json: cannot unmarshal object into Go value of type []atc.Version

when parsing resource response:

{
  "version": {
    "ref": "61cebf"
  },
  "metadata": [
    {
      "name": "commit",
      "value": "61cebf"
    },
    {
      "name": "author",
      "value": "Hulk Hogan"
    }
  ]
}
```
wrapping it in an array changes the error:

```
run check step: run check step: check: json: cannot unmarshal object into Go value of type string

when parsing resource response:

[
  {
    "version": {
      "ref": "61cebf"
    },
    "metadata": [
      {
        "name": "commit",
        "value": "61cebf"
      },
      {
        "name": "author",
        "value": "Hulk Hogan"
      }
    ]
  }
]
```

but how is that response structure supposed to be?

I do see the [git-resource](https://github.com/concourse/git-resource) and also docker-image resource do this, I just don't get **how** they do it

**Comment by taylorsilva** [comment, marked as answer]
It looks like you mixed up your `check` and `in` responses. What your `check` is currently returning is the response Concourse expects from the `in` script, not the `check` script.

Your check response should be in this format, an array of json objects: 
```
[
  { ... }
  { ... }
  { ... }
]
```

But currently you have it in this format, which is the expected response from the `in` script, **not** the `check` script.
```
{
  "version": {}
  "metadata: [ {}, {}...]
}
```

Hope that clears things up for you.

**Reply by x3nb63** [reply]
> Hope that clears things up for you.

not yet, but thanks for responding.

where would I place "metadata" in this format?

> ```
> [
>   { ... }
>   { ... }
>   { ... }
> ]
> ```

when I look at a git-resource at the Web UI, I see a list of `ref abc...` versions it knows. When I click the first one, it shows a table with three columns: _inputs to_, _outputs of_ and _metadata_.

this third column has additional information such as author, author_date, ..., branch, message ...

this metadata is what I want my resource to show too, because - quite similar to git - my version is a UUID which tells nothing to the human.

**Reply by taylorsilva** [reply]
> where would I place "metadata" in this format?

The `check` script doesn't have a spot to return metadata. It only returns versions, not metadata about the version.

Metadata is returned by the `in` script which is run by [`get` steps](https://concourse-ci.org/jobs.html#get-step) in your jobs. The `in` script has a `metadata` key where metadata is then rendered
```
{
  "version": {}
  "metadata: [ {}, {}...]
}
```

Maybe this visual helps. I think you're a bit lost about when and where things are getting run.
<img width="994" alt="image" src="https://user-images.githubusercontent.com/4583837/106314727-272ec800-6238-11eb-9abd-e363720f5fa2.png">

More clarifying info to help you.

The `check` script is expected to return **versions** of something. A "version" has to be uniquely identifying so the check script should only return that info. Metadata is not unique (e.g. author can be the same for many different commits).

The `in` script (run by `get` steps) will take in a single version that was produced by the `check` script (usually the latest version returned by `check`) and return the entirety of whatever represents that version. If that involves files then the `in` script can fetch those files and stores them in the path passed in as the first argument to the `in` script. On stdout the `in` script can also return `metadata` about the version in this format.
```
{
  "version": {}
  "metadata: [ {}, {}...]
}
```

**Reply by x3nb63** [reply]
ok, so I understand the WebUI - on the resource page - prints the metadata returned by the corresponding `in` (get) step of the particular job build number.

So that `metadata` column shows additional info of the build referenced in the _inputs to_ column, right?

In that case I know what to do, probably.

thanks again.

**Reply by taylorsilva** [reply]
Yup you got it! The metadata will only show on the resource page once a `get` step runs and returns any metadata.
